lib/src/zenohUri.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2024 General Motors GTO LLC
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2024 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+ 
+#include <iomanip>
+#include <sstream>
+#include <spdlog/spdlog.h>
+#include <up-cpp/uri/serializer/MicroUriSerializer.h>
+#include "up-client-zenoh-cpp/session/zenohSessionManager.h"
+
+namespace uprotocol::uri {
+
+namespace {
+void hexlify(auto& topic, const uint8_t byte) {
+    topic << std::hex << std::setw(2) << std::setfill('0');
+    topic << static_cast<int>(byte);
+}
+
+void hexlify(auto& topic, const auto &&start, const auto &&end) {
+    for (auto i = start; i < end; ++i) {
+        hexlify(topic, *i);
+    }
+}
+
+} // anonymous namespace
+
+std::string toZenohKeyString(const uprotocol::v1::UUri &u_uri) {
+    constexpr std::string_view remote_prefix{"upr"};
+    constexpr std::string_view local_prefix{"upl"};
+    constexpr std::string_view separator{"/"};
+    constexpr std::string_view wildcard_suffix{"**"};
+
+    using Serializer = uprotocol::uri::MicroUriSerializer;
+    constexpr auto authority_offset = Serializer::AuthorityStartPosition;
+
+    std::ostringstream topic;
+
+    if (u_uri.has_authority() && !u_uri.has_entity() && !u_uri.has_resource()) {
```

Comment:
as this code will be called each time send is called, do you know if there is any performance impact?

Reply:
If you're asking specifically about line 58, then no I am not worried about the performance impact. It generates 11 instructions on x86-64 and 16 instructions on arm64 (with [gcc 11.4 and the `-O3` flag](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAOykrgAyeAyYAHI%2BAEaYxCBmAGykAA6oCoRODB7evgGp6ZkCoeFRLLHxSbaY9o4CQgRMxAQ5Pn6BdpgOWQ1NBCWRMXEJyQqNza15HeP9YYPlw0kAlLaoXsTI7BzmAMxhyN5YANQmO27IY/iCp9gmGgCCd/djxF4OR3L3XgQIJIQAnid/FYHkcwUcAG6oPDoABUR3QTEapxBj38ABEUU8ni83gQPthBACgajwZDoXCEUimFiHiYMbTHg9ce85AAlTAKNYbTAkp5kqEw%2BGI5E7VH0zFi7HMgivVlyYh4Pmg8FhWhzI7RVCeI4IJgKAD6TG%2Bv0VBH%2BECWRzQDDGyvuZLJxEwBHWDCOBrCBDizFoBr1huNPz%2B5stjLJEpOdJVYLVGq1OoDBsYjlDVptdvppMdYOdruI7s9gh9Yn9%2BuTRLT4fBkelDtVDHV4U12touvLzq5602lutAkzwP5OaOebdHq9Jb9Sc73J7S2rYNr0fuKUVEKR7CHscb8db4%2BLBdLSaDpoBvYz%2BKzW8do4LRx%2BeAURzAYFO6KOE8PfqwVGNtAIRZjIYc4nEk5iJEcJ4hv8BrPq%2BOzvgwXi0LQKSyguQLvnWZJxs2CZtkW3pfmWhopme6b9peg4xjeLpjg%2BT4vm%2BH4Hr6Bo/n%2BAFhEBDAgeB4FHGR5qwUxCFHEhKFocQGFLky9bbk2vL4fuRFsdOnKzpg56Ufaw4jnRd4MXBzGfmxHHIVxtqNLxWlWvxST6V2PIifBiHIah6FSjRsnXkBjjICxil9lZHwKng8KmaW5n/oB1l8dR8m0fm7rOp%2Bq4ugaohjKc5w6XIYWwjcEDgQa%2BWKuxmC/hZsXAZgBrzl5iUSth4KfCa0HwlBZowRhciEqm/zwkJPWNWS7Iad2mDwjOk0Ghhfl4AFF6hYqHplXgFVVTF3FxXVjLNY1TzrcFdrrSAIClWFm2cTVNlzYODKHQ8ynIAgXQANYQMdF4CV4BrrHgdkJU6BnuhAf0AwAdMe7Xdb29kQS%2BEOKtD5bDfDYEOUj/0o%2BpTlzg14oYhwKy0JwACsvB%2BBwWikKgnBuNY1hHPjvK7DwpAEJoJMrO9IDkxo%2BicJIVPc3TnC8AoICC1zNMk6QcCwDAiAoKgLApHQcTkJQaDq5r8TAFwUh8HQRFSxA0Ri9EYRNP8nAc9bzDEP8ADy0TaF0ssc7rbCCC7jZ23LpBYNEXjAG4Yi0FL3C8FgLCGMA4hB/gzrdBCnJi5gqhdN8WwcxOZNB%2Bq0TELbHhYGLsp4Cw9u8OnxBako6KYPHRhNqAcsrFQBjAAoABqeCYAA7i7KSMLXMiCCIYjsMb/CCIoKjqEHuhcPoCcoEzlj6Hg0RS5AKyoGhWTRwAtC7Oy8Kg9eKlg%2B%2BWtUtRZC4DDuJ4bQSAAnEEr8DGUFQSGSGkDIdRsjvzyFwb%2BwCigMD/kMeIXBRg1E9j0GYkw/CQMfig%2BoMw4ELAQaMNB4CMHfyAs0PBADEErFZhIUmFNRZB3phwI4qgAAciRT6JEkEcYAyAApG0htwiAuBCAkFAjsLgSxeCyy0EsXm/NBaFxFqQGuAtSDU1pkwyW0tObcxWIrFWusNb0DIBQCARj9YgCaCwLhhspCCxoP%2BOI5tLZB0drbCe7jnZuw9g4CePsUz%2B1oIHWmIcw4RxQtHDmccE5J1CXgVOjh07R1plnHO3oJ4FzFsXUuzty5bFplXGuMdSD10bpgZurdE5hA7rIvgPd%2B6DxHmPamHN57CFEOIOeshF5qDFroMw68jCb0sNYHee94CH2Pv2Tg59L50xvjCDOB8sHdGcBAVw6C9AhDmP/YYa9oGgK2QcwooCKH7NWaA3oExiF6E6Gshg1zZilHgXcohuQMG2Fwbs15kjViTVoULDglN1FiyYUce4bIACyXCeF8KOAIyQkMNBHGEfgIgxBxGSOkXolYb0mBYHiA/JRvBVGCw0VfCWtgdEyJ5qQPmajC7zIpeLDgOLO50I4GYBhmiqW0rkaU5xz9JBAA) set, based on the code protobuf generates).

If you mean in general calling the serializer then converting the byte array to hexadecimal, then yeah I am a little worried. However, this is (based on a quick skim) equivalent to the rust implementation right now, and will get somewhat smaller when we switch to short URIs.

We could mitigate this some in the future by using some other easy to compute key in our maps, then attaching to those the pre-converted URI string. I think it would be worth measuring and optimizing after we are done with the 1.5.7 updates.

Reply:
Yes , i am talking about the whole conversion function :) 
One possible way to address this , is by using a MAP , that will contain the already calculated values for the specific URI , this way you will do it only once.

Can you please do a simple check and see how much time i takes to calculate one URI?

Reply:
@MelamudMichael - This code is going to be rewritten in 1.5.8 with the change to short-form URIs, so it's probably best to do that performance evaluation and consider improvements at that time. I've created #29 to track it. It would also probably be worth doing a general sweep to identify other critical paths for optimization.